File: tasks/letters_counter.rb
def process str
    current_letter = str[0]
    counter = 1
    result = ''
  
    1.upto str.length do |i|
      current_char = str[i]
  
      if current_char == current_letter
        counter += 1
      else
        if counter > 1
          result = "#{result}#{counter}#{current_letter}"
        else
          result = "#{result}#{current_letter}"
        end
        counter = 1
        current_letter = current_char
      end
  
    end
  
    result
  end
  
  puts process 'aabcdefffg' # 2abcde3fg
  puts process 'aaabbbbcccccdddddd' # 3a4b5c6d
  puts process 'adkjlnnnabbatrsfaaag' # adkjl3na2batrsf3ag
  puts process 'abc' # abc
  puts process 'a' # a

File: tasks/letters_counter.ts
// letter counter 'aabbcc' => '2a2b2c'

const process = (str: string): string => {
    let result = '';
    let track = str[0];
  
    for (let i = 1; i <= str.length; i++) {
      const letter = track[0];
  
      if (str[i] !== letter) {
        const currentResult = track.length > 1 ? `${track.length}${letter}` : `${track}`;
        result = `${result}${currentResult}`;
        track = str[i];
      } else {
        track = `${track}${str[i]}`;
      }
    }
  
    return result;
  }
  
  console.log(
    process('aabcdefffg') // 2abcde3fg
  );
  
  console.log(
    process('aaabbbbcccccdddddd') // 3a4b5c6d
  );
  
  console.log(
    process('adkjlnnnabbatrsfaaag') // adkjl3na2batrsf3ag
  );
  
  console.log(
    process('abc') // abc
  );
  
  console.log(
    process('a') // a
  );